backend/inspirehep/pidstore/models.py

```diff
@@ -0,0 +1,139 @@
+            "pidstore_pid.id", name="fk_inspire_pidstore_redirect_new_pid_id"
+        ),
+        nullable=False,
+        index=True,
+    )
+
+    new_pid = db.relationship(
+        "PersistentIdentifier",
+        backref="redirected_pids",
+        foreign_keys=new_pid_id,
+        innerjoin=True,
+        uselist=False,
+    )
+
+    @classmethod
+    def _fix_existing_redirection(cls, redirection, old_pid, new_pid):
+        if old_pid.status != PIDStatus.REDIRECTED:
+            raise PidStatusBroken(old_pid)
+
+        elif (
+            redirection.new_pid.pid_type == new_pid.pid_type
+            and redirection.new_pid.pid_value == new_pid.pid_value
+        ):
+            LOGGER.info(
+                "Pid already redirected correctly.", old_pid=old_pid, new_pid=new_pid
+            )
+        else:
+            redirection.new_pid = new_pid
+            db.session.add(redirection)
+        return redirection
+
+    @classmethod
+    def _create_new_redirection(cls, old_pid, new_pid):
+        try:
+            with db.session.begin_nested():
+                redirection = cls(original_pid=old_pid, new_pid=new_pid)
+                db.session.add(redirection)
+                old_pid.status = PIDStatus.REDIRECTED
+                db.session.add(old_pid)
+        except IntegrityError as e:
+            raise PIDInvalidAction(e)
+        except SQLAlchemyError as e:
+            LOGGER.exception(
+                "Failed to redirect record", old_pid=old_pid, new_pid=new_pid
+            )
+            raise
+        return redirection
+
+    @classmethod
+    def redirect(cls, old_pid, new_pid):
+        """Redirects pid from old_pid to new_pid"""
+        if old_pid.pid_type != new_pid.pid_type:
+            raise WrongPidTypeRedirection(old_pid, new_pid)
+
+        if (
+            not old_pid.is_registered()
+            and not old_pid.is_deleted()
+            and not old_pid.is_redirected()
+        ) or not new_pid.is_registered():
+            raise WrongRedirectionPidStatus(old_pid, new_pid)
+
+        try:
+            redirection = cls.get(old_pid, reload=True)
+            redirection = cls._fix_existing_redirection(redirection, old_pid, new_pid)
+        except PidRedirectionMissing:
+            redirection = cls._create_new_redirection(old_pid, new_pid)
+
+        LOGGER.info("PID redirected successfully", old_pid=old_pid, new_pid=new_pid)
+        return redirection
+
+    @classmethod
+    def get_redirect(cls, pid):
+        while pid.status == PIDStatus.REDIRECTED:
+            pid = cls.get(pid).new_pid
+        return pid
+
+    @classmethod
+    def get(cls, original_pid, reload=False):
+        redirection = None
+        if reload:
+            redirection = cls.query.filter_by(original_pid=original_pid).one_or_none()
+        elif len(original_pid.redirection) == 1:
+            redirection = original_pid.redirection[0]
+        if not redirection:
+            raise PidRedirectionMissing(original_pid)
```

Comment:
why not simply return `None`?

Reply:
It's done like that because Invenio would not know what to do with `None` when it's trying to resolve a PID.

Reply:
But what does it have to do with that? AFAICT, it's used only in your code, not in the pid provider.

Reply:
https://github.com/inspirehep/invenio-records-rest/blob/f398bab4f4f96607045e753d25aa09bb404e9e42/invenio_records_rest/utils.py#L154 is calling resolver which is throwing Exception if record is redirected with new pid as a parameter to this exception https://github.com/inspirehep/inspirehep/pull/1506/files#diff-bab823ce6da991443e3f70c6c8dd83b71bd9cf306f65979d2f4278c4234330e9R45 So, if get() would not throw exception when redirection is missing then I would have to manually trigger this exception in get_redirect or in resolve, otherwise Invenio would not know what to do with `None` pid.
Also it's expected usually that if method `get` didn't found an object then it's throwing some exception.